src/app/rpmostree-builtin-start-daemon.cxx

```diff
@@ -46,12 +46,9 @@ typedef enum {
 static AppState appstate = APPSTATE_STARTING;
 static gboolean opt_debug = FALSE;
 static char *opt_sysroot = NULL;
-static gint service_dbus_fd = -1;
 static GOptionEntry opt_entries[] =
 {
   { "debug", 'd', 0, G_OPTION_ARG_NONE, &opt_debug, "Print debug information on stderr", NULL },
-  { "sysroot", 0, 0, G_OPTION_ARG_STRING, &opt_sysroot, "Use system root SYSROOT (default: /)", "SYSROOT" },
```

Comment:
Can we leave this one in? It's related to the peer D-Bus stuff, but still mostly independent, right?

There's a bunch of things related to this I'd like to explore. E.g. running `rpm-ostree` directly after `coreos-installer install`, running `rpm-ostree` from the initramfs, etc... I think also Anaconda may need this for the `initramfs-etc` work.

Reply:
Cool, this actually works today from a live boot:

```
[root@cosa-devsh ~]# coreos-installer install /dev/vda
[root@cosa-devsh ~]# mount /dev/disk/by-label/root /mnt/root
[root@cosa-devsh ~]# mount /dev/disk/by-label/boot /mnt/root/boot
[root@cosa-devsh ~]# rpm-ostree start-daemon --sysroot /mnt/root &
[root@cosa-devsh ~]# rpm-ostree status
State: idle
Deployments:
* ostree://fedora:fedora/x86_64/coreos/testing-devel
                   Version: 34.20210520.dev.0 (2021-05-20T21:45:14Z)
                    Commit: a53b7d233ef46c2e1ceb469c0ec8e441a5d67c03c6641630632e89c52f45a516
              GPGSignature: (unsigned)
```

We'd want to make that easier of course. Maybe just have coreos-installer automatically add a drop-in for `rpm-ostreed.service` or something. Not sure.

Actually doing a write operation like `rpm-ostree ex initramfs-etc` on it fails, but at least the base is there.

Reply:
> E.g. running rpm-ostree directly after coreos-installer install,

But the more we do that the more it weakens the Ignition story/focus.

>  running rpm-ostree from the initramfs, etc... 

I am very skeptical of this one because any nontrivial use will need to *configure* rpm-ostree (think e.g. 3rd party repos or local mirroring) and then that gets into configuring the initramfs and...

I think most of the desire for rpm-ostree-in-initramfs has been obviated by the `install -A` model.

Also we don't have tests for `--sysroot` now.  It might not be really hard with some scaffolding to to basically double our tests and do it though.  Or I guess have some basic tests.

Reply:
> > E.g. running rpm-ostree directly after coreos-installer install,
> 
> But the more we do that the more it weakens the Ignition story/focus.

Oh yes definitely. But I'm hoping to converge towards `rpm-ostree ex initramfs-etc` for solving bootstrapping problems which hard require pre-configuration. Notably networking and non-default multipath configuration.

> > running rpm-ostree from the initramfs, etc...
> 
> I am very skeptical of this one because any nontrivial use will need to _configure_ rpm-ostree (think e.g. 3rd party repos or local mirroring) and then that gets into configuring the initramfs and...
>
> I think most of the desire for rpm-ostree-in-initramfs has been obviated by the `install -A` model.

I haven't thought this through yet (and I've actually been meaning to experiment with this during a hack day), but as part of https://github.com/coreos/fedora-coreos-tracker/issues/681, basically I was thinking we would match Ignition semantics and apply layered requests *before* we switchroot. I think this would be cleaner just doing it in a bwrap container though to avoid actually having to drag rpm-ostree into the initrd. So maybe we don't actually need `--sysroot` in that case. But yeah, configuration would definitely come from the real root.

The advantage over live-apply in the real root is that it removes any ambiguity wrt ordering and e.g. configuring the layered things via Ignition itself. It also allows us to enforce strict failure semantics like Ignition.

> Also we don't have tests for `--sysroot` now. It might not be really hard with some scaffolding to to basically double our tests and do it though. Or I guess have some basic tests.

Yeah, I think some basic tests at least wouldn't be too hard. I can volunteer for that! :)

Reply:
OK for now, I kept sysroot since it is orthogonal to removing the session bus bits.

That said, on this topic...my thought here is that we can avoid using DBus for these cases.  Basically `rpm-ostree ex initramfs-etc --sysroot=/mnt/sysroot ...` would directly perform the operation in the client code.

It's much like where systemd ended up where e.g. `systemctl` can talk to pid1 over dbus, but also happily works offline directly.

Reply:
> That said, on this topic...my thought here is that we can avoid using DBus for these cases. Basically rpm-ostree ex initramfs-etc --sysroot=/mnt/sysroot ... would directly perform the operation in the client code.

I think that makes sense though likely will require a lot of refactoring to pull off. Definitely worth investigating though!